tests/mir_test_framework/canonical_window_manager.cpp

```diff
@@ -418,21 +421,30 @@ void msh::CanonicalWindowManagerPolicy::handle_delete_surface(std::shared_ptr<ms
     fullscreen_surfaces.erase(surface);
     bool const is_active_surface{surface.lock() == active_surface()};
 
-    auto& info = tools->info_for(surface);
-
-    if (auto const parent = info.parent.lock())
+    try
     {
-        auto& siblings = tools->info_for(parent).children;
+        auto& info = tools->info_for(surface);
 
-        for (auto i = begin(siblings); i != end(siblings); ++i)
+        if (auto const parent = info.parent.lock())
         {
-            if (surface.lock() == i->lock())
+            auto& siblings = tools->info_for(parent).children;
+
+            for (auto i = begin(siblings); i != end(siblings); ++i)
             {
-                siblings.erase(i);
-                break;
+                if (surface.lock() == i->lock())
+                {
+                    siblings.erase(i);
+                    break;
+                }
             }
         }
     }
+    catch (std::out_of_range e)
+    {
+        // tools->info_for(surface) throws std::out_of_range if the surface isn't in a map
+        // so far I only know of this happening in shutdown after WLCS XDG shell popup tests
+        log_warning("std::out_of_range caught in msh::CanonicalWindowManagerPolicy::handle_delete_surface()");
```

Comment:
How does this get called without the surface being in the map? Is this a strange behaviour that would also bite other window manager implementations?

Reply:
tests/mir_test_framework/canonical_window_manager.cpp is code we need to replace. I know it has a bunch of consistency issues. I once tried to rip it out only to find lots of tests rely on its idiosyncratic behaviour. (That's why it is under "tests" not under "src".)

It is unfortunate that wlcs built on this shaky foundation.

Reply:
Specifically, the issue arises when the surface is a popup. My best guess is the map is for real windows, so popups don't get added.

Reply:
My best guess is that mir_test_framework/canonical_window_manager.cpp has not registered (or has already de-registered) the window because it was written long before we added the Wayland frontend and events don't always happen in the same order they did with libmirclient. (We had to fixed some issues like this in the libmiral window manager.)

If you could share how to produce the error you are seeing we could confirm that theory.

Naturally, the best answer is to move wlcs away from using this obsolete code.